modules/gapi/src/backends/oak/goakbackend.cpp

```diff
@@ -0,0 +1,597 @@
+        construct_impl(ctx,
+                       node,
+                       in_queues_params,
+                       typename cv::detail::MkSeq<sizeof...(Ins)>::type(),
+                       typename cv::detail::MkSeq<sizeof...(Outs)>::type());
+    }
+};
+
+} // namespace detail
+
+struct GOAKKernel {
+    using F = std::function<void(GOAKContext&,
+                                 std::shared_ptr<dai::Node>&,
+                                 std::vector<std::pair<std::string, dai::Buffer>>&)>;
+    explicit GOAKKernel(const F& f) : m_f(f) {}
+    const F m_f;
+};
+
+struct OAKComponent
+{
+    static const char *name() { return "OAK Component"; }
+    GOAKKernel k;
+};
+
+}} // namespace gimpl // namespace cv
+
+using OAKGraph = ade::TypedGraph
+    < cv::gimpl::OAKComponent
+    // FIXME: extend
+    >;
+
+using ConstOAKGraph = ade::ConstTypedGraph
+    < cv::gimpl::OAKComponent
+    // FIXME: extend
+    >;
+
+void cv::gimpl::GOAKExecutable::LinkToParentHelper(ade::NodeHandle handle,
+                                                   const std::vector<ade::NodeHandle>& nodes)
+{
+    ade::NodeHandle parent;
+    for (const auto& indatah : handle.get()->inNodes()) {
+        // indatah - node's input data
+        // need to find which other node produces that data
+        for (const auto& nhp : nodes) {
+            if (m_gm.metadata(nhp).get<NodeType>().t == NodeType::OP) {
+                for (const auto& outdatah : nhp.get()->outNodes()) {
+                    if (indatah == outdatah) {
+                        parent = nhp;
+                    }
+                }
+            }
+        }
+    }
+    GAPI_Assert(m_oak_nodes.find(parent) != m_oak_nodes.end());
+    GAPI_Assert(m_oak_node_inputs[handle].size() ==
+                m_oak_node_outputs[parent].size());
+    for (size_t i = 0; i < m_oak_node_inputs[handle].size(); ++i) {
+        m_oak_node_outputs[parent][i]->link(*(m_oak_node_inputs[handle][i]));
+    }
+}
+
+cv::GArg
+cv::gimpl::GOAKExecutable::packArg(const GArg &arg,
+                                   std::vector<ExtractTypeHelper::InputPtr>& oak_ins,
+                                   std::vector<cv::GArg>& oak_wrapped_args) {
+    if (arg.kind != cv::detail::ArgKind::GOBJREF) {
+        GAPI_Assert(   arg.kind != cv::detail::ArgKind::GMAT
+                    && arg.kind != cv::detail::ArgKind::GSCALAR
+                    && arg.kind != cv::detail::ArgKind::GARRAY
+                    && arg.kind != cv::detail::ArgKind::GOPAQUE
+                    && arg.kind != cv::detail::ArgKind::GFRAME);
+        // All other cases - pass as-is, with no transformations to
+        // GArg contents.
+        return arg;
+    }
+    const cv::gimpl::RcDesc &ref = arg.get<cv::gimpl::RcDesc>();
+    switch (ref.shape) {
+    case GShape::GFRAME:
+        oak_ins.push_back(nullptr);
+        oak_wrapped_args.push_back(GArg(&(oak_ins.back())));
```

Comment:
pointer on nullptr looks confusing

If this packArg is one time shot method- it must be surrounded by GAssert at least.
Otherwise the subsequent call of packArg can invalidate oak_ins continuous memory location and oak_wrapped_args would become dangling

Reply:
Reserved oak_wrapped_args according to the total number of wrapped arguments in OAK subgraph

Reply:
at least comment are required, or big-thick method which called `reservation_initial_param_to_avoid_reallocation_and_dangling_pointers_problem` would be better.

i asked, because in case of some trouble the first things which would be under investigation after about 3~4 month in case of any issue is memory-allocation here. To avoid waste time on that investigation - it would recommended to add some kind of protection/or clearest logic here


P.S. whats about  `std::shared_ptr` in case you do not copy Input & Outputs ?